libraries/hwlib/library/targets/usart_queue.hpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * =====================================================================================
+ *
+ *       Filename:  usart_queue.hpp
+ *
+ *    Description:  
+ *
+ *        Version:  1.0
+ *        Created:  14-03-19 11:25:21
+ *       Revision:  none
+ *       Compiler:  gcc
+ *
+ *         Author:  YOUR NAME (), 
+ *   Organization:  
+ *
+ * =====================================================================================
+ */
+#ifndef QUEUE_HPP
+#define QUEUE_HPP
+#include "hwlib.hpp"
+
+template <class T, size_t QUEUE_SIZE>
+class Queue {
+private:
+	unsigned int _front, _back, _count;
+	T _data[QUEUE_SIZE];
+
+public:
+	Queue() {
+		_front = 0;
+		_back = 0;
+		_count = 0;
+	}
+	inline int count();
+	inline int front();
+	inline int back();
+	void push(const T &item);
+	T peek();
+	T pop();
+	void clear();
+};
+
+template <class T, size_t QUEUE_SIZE>
+inline int Queue<T, QUEUE_SIZE>::count() {
+	return _count;
+}
+
+template <class T, size_t QUEUE_SIZE>
+inline int Queue<T, QUEUE_SIZE>::front() {
+	return _front;
+}
+
+template <class T, size_t QUEUE_SIZE>
+inline int Queue<T, QUEUE_SIZE>::back() {
+	return _back;
+}
+
+template <class T, size_t QUEUE_SIZE>
+void Queue<T, QUEUE_SIZE>::push(const T &item) {
+	if (_count < QUEUE_SIZE) { // Drops out when full
+		_data[_back++] = item;
+		++_count;
+		// Check wrap around
+		if (_back > QUEUE_SIZE)
+			_back -= (QUEUE_SIZE + 1);
+	}
+}
+
+template <class T, size_t QUEUE_SIZE>
+T Queue<T, QUEUE_SIZE>::pop() {
+	if (_count <= 0)
```

Comment:
why does this check if the _count is below or equal zero? _count is a unsigned int so it can't go below zero.